clang/lib/Sema/SemaTemplateInstantiateDecl.cpp

```diff
@@ -1614,14 +1614,38 @@ Decl *TemplateDeclInstantiator::VisitVarDecl(VarDecl *D,
   // Only add this if we aren't instantiating a variable template.  We'll end up
   // adding the VarTemplateSpecializationDecl later.
   if (!InstantiatingVarTemplate) {
-    SemaRef.addSyclVarDecl(Var);
+    if (SemaRef.getLangOpts().SYCLIsDevice &&
+        SemaRef.isTypeDecoratedWithDeclAttribute<SYCLDeviceGlobalAttr>(
+            Var->getType())) {
+      if (!Var->hasGlobalStorage())
+        SemaRef.Diag(D->getLocation(),
+                     diag::err_sycl_device_global_incorrect_scope);
+
+      if (Var->getAccess() != AS_public)
```

Comment:
@premanandrao this is likely not the right location to do that because  https://github.com/intel/llvm/pull/5810#discussion_r956207644 
This place also covers cases where device_global is instantiated through some method calls, where the access specifier turns out to be AS_none (where there's no explicit access specifier such as in structs/classes), but they should be diagnosed because they are not publicly accessible.

Reply:
> @premanandrao this is likely not the right location to do that because [#5810 (comment)](https://github.com/intel/llvm/pull/5810#discussion_r956207644) This place also covers cases where device_global is instantiated through some method calls, where the access specifier turns out to be AS_none (where there's no explicit access specifier such as in structs/classes), but they should be diagnosed because they are not publicly accessible.

I saw that comment, but I am not quite sure I understand.  Why do you want to diagnose AS_none cases when they are enclosed in a struct (and have default public access)?  What's a specific example?

Reply:
I meant this case from L71 of the test. 
Var->getAccess() or D->getAccess() returns AS_none

```
template <typename T> void fooBar() {
  // expected-error@+2{{'device_global' member variable 'c' should be publicly accessible from namespace scope}}
  // expected-error@+1{{'device_global' variable must be a static data member or declared in global or namespace scope}}
  static device_global<T> c;
  // expected-error@+2{{'device_global' member variable 'd' should be publicly accessible from namespace scope}}
  // expected-error@+1{{'device_global' variable must be a static data member or declared in global or namespace scope}}
  device_global<T> d;
}

int main() {
  // expected-note@+1{{in instantiation of function template specialization 'fooBar<int>' requested here}}
  fooBar<int>();
}
```

Reply:
But in this case, the variable 'c' or 'd' is local to the function 'fooBar'; it is not a struct or a class data member.  There is no public/protected/private access.

Reply:
@schittir does deleting this check cause any tests to fail? And if so, should the context be checked here instead of access?

Reply:
@premanandrao - Are you saying that it is not appropriate to diagnose this specific case based on access specifier at all, because it is not a struct or class? 
If yes, I can simply change the check to `if (Var->getAccess() == AS_private || Var->getAccess() == AS_protected)`, and all other cases are covered/unimpacted. 
@elizabethandrews What do you think?

Reply:
Changed it per Prem's suggestion because it is unnecessary to diagnose this case with access specifier https://github.com/intel/llvm/pull/5810#discussion_r957791510

Reply:
This particular place not only handled the case above but also member variable instantiations. Sorry about the confusion here https://github.com/intel/llvm/pull/5810#discussion_r957612695 
I fixed it now.